library/Icinga/Application/Hook/DBMigrationHook.php

```diff
@@ -0,0 +1,164 @@
+<?php
+
+/* Icinga Web 2 | (c) 2023 Icinga GmbH | GPLv2+ */
+
+namespace Icinga\Application\Hook;
+
+use DirectoryIterator;
+use Icinga\Application\Hook\Common\DBMigration;
+use Icinga\Application\Icinga;
+use Icinga\Application\Modules\Module;
+use SplFileInfo;
+
+/**
+ * @inheritDoc
+ *
+ * The version numbers of the sql migrations are determined by extracting the respective migration script names.
+ * It's required to place the sql migrate scripts below the respective following directories:
+ *
+ *   `{IcingaApp,Module}::baseDir()/schema/{mysql,pgsql}-upgrades`
+ */
+abstract class DBMigrationHook extends MigrationHook
+{
+    public const MYSQL_UPGRADE_DIR = 'schema/mysql-upgrades';
+
+    public const PGSQL_UPGRADE_DIR = 'schema/pgsql-upgrades';
+
+    /** @var array<string, DBMigration> All pending DB migrations of this hook */
+    protected $migrations;
+
+    /**
+     * Get the migration count of this hook
+     *
+     * @return int
+     */
+    public function count(): int
+    {
+        return count($this->migrations);
+    }
+
+    /**
+     * Get all the pending migrations of this hook
+     *
+     * @return DBMigration[]
+     */
+    public function getMigrations(): array
+    {
+        return $this->migrations;
+    }
+
+    public function apply(): bool
+    {
+        if (! $this->preApply()) {
+            return false;
+        }
+
+        foreach ($this->getMigrations() as $migration) {
+            if (! $this->applyOne($migration->getVersion())) {
```

Comment:
This also runs `preApply` again. If that's unavoidable, `preApply` should at least cache its result.

Reply:
It must be checked whenever you try to run it, since it is associated with other hooks, i.e. if `preApply()` returns false now, it might return true after applying some Icinga Web 2 hooks.